src/rust/catnap/mod.rs

```diff
@@ -140,25 +160,29 @@ impl CatnapLibOS {
         trace!("bind() qd={:?}, local={:?}", qd, local);
 
         // Issue bind operation.
-        match self.sockets.get(&qd) {
-            Some(&fd) => {
-                let sockaddr: libc::sockaddr_in = linux::socketaddrv4_to_sockaddr_in(&local);
-                match unsafe {
-                    libc::bind(
-                        fd,
-                        (&sockaddr as *const libc::sockaddr_in) as *const libc::sockaddr,
-                        mem::size_of_val(&sockaddr) as u32,
-                    )
-                } {
-                    stats if stats == 0 => Ok(()),
-                    _ => {
-                        let errno: libc::c_int = unsafe { *libc::__errno_location() };
-                        error!("failed to bind socket (errno={:?})", errno);
-                        Err(Fail::new(errno, "operation failed"))
-                    },
-                }
+        match self.qtable.get(&qd) {
+            Some(queue) => match queue.get_fd() {
+                Some(fd) => {
+                    let sockaddr: libc::sockaddr_in = linux::socketaddrv4_to_sockaddr_in(&local);
+                    match unsafe {
+                        libc::bind(
+                            fd,
+                            (&sockaddr as *const libc::sockaddr_in) as *const libc::sockaddr,
+                            mem::size_of_val(&sockaddr) as u32,
+                        )
+                    } {
+                        stats if stats == 0 => Ok(()),
+                        _ => {
+                            let errno: libc::c_int = unsafe { *libc::__errno_location() };
+                            error!("failed to bind socket (errno={:?})", errno);
+                            Err(Fail::new(errno, "operation failed"))
+                        },
+                    }
+                },
+                None => Err(Fail::new(libc::EBADF, "invalid file descriptor")),
             },
-            _ => Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
+            // TODO: Return the queue descriptor
```

Comment:
I did not get this. Why should we return the queue descriptor if we fail?

Reply:
I should have said, print the queue descriptor. We have many of these checks but most of them do not print the queue descriptor in the error.